head.go

```diff
@@ -601,9 +601,112 @@ func (h *Head) Delete(mint, maxt int64, ms ...labels.Matcher) error {
 	for _, s := range stones {
 		h.tombstones.add(s.ref, s.intervals[0])
 	}
+
+	_, err = h.CleanTombstones()
+	if err != nil {
+		return errors.Wrap(err, "Head: CleanTombstones in Delete")
+	}
+
 	return nil
 }
 
+// CleanTombstones re-writes the chunks with tombstones
+// and deletes chunks if there is no samples after removing tombstones 
+func (h *Head) CleanTombstones() (bool, error) {
+
+	modified := false // flag for entire block
+
+	for sid, dranges := range h.tombstones {
+		if len(dranges) == 0 {
+			continue
+		}
+
+		cleaned := false // flag for current tombstone
+		stripeSeriesIndex := sid%stripeSize
+
+		h.series.locks[stripeSeriesIndex].Lock()
+		s := h.series.series[stripeSeriesIndex] // series with tombstones
+
+		if ms, ok := s[sid]; ok {
+			// Deleting samples for intervals in tombstones
+			for i, chk := range ms.chunks {
+				if !intervalOverlap(dranges[0].Mint, dranges[len(dranges)-1].Maxt, chk.minTime, chk.maxTime) {
+					continue
+				}
+
+				// create new chunk by removing deleted samples
+				newChunk := chunkenc.NewXORChunk()
+				app, err := newChunk.Appender()
+				if err != nil {
+					return modified, err
+				}
+
+				modified = true
+				cleaned = true
+
+				it := &deletedIterator{it: chk.chunk.Iterator(), intervals: dranges}
+				for it.Next() {
+					ts, v := it.At()
+					app.Append(ts, v)
+				}
+
+				if newChunk.NumSamples() > 0 {
+					ms.chunks[i].chunk = newChunk
+				} else { // empty chunk, no required to store
+					ms.chunks = append(ms.chunks[:i], ms.chunks[i+1:]...)
```

Comment:
There is a race here? TODO(gouthamve): Explain with golang playground.

Reply:
Series is locked [here at line 627](https://github.com/prometheus/tsdb/pull/270/files#diff-0ced3454d447442a96a77b33ebe87300R627). Those operations are done before unlocking. So I think there is no race here. Or am I missing something?